server/bin/pbench-audit-server.py

```diff
@@ -0,0 +1,581 @@
+    return 0
+
+
+# Controller Hierarchy
+def verify_controllers(hier):
+    """ 'hier' is a ControllerHierarchy class object. """
+    controllers = Path(hier.path).glob("*")
+    verify_valid_controllers(hier, controllers)
+
+    for controller in hier.controllers:
+        dir_path = Path(hier.path, controller)
+        unexpected_dirs = list()
+        if not Path(hier.config.ARCHIVE, controller).is_dir():
+            """We have a controller in the hierarchy which does not have a
+            controller of the same name in the archive hierarchy.  All
+            we do is report it, don't bother analyzing it further.
+            """
+            hier.add_controller_list(Hierarchy.MIALIST, Path(controller).name)
+        else:
+            """Report any controllers with objects other than directories
+            and links, while also recording any empty controllers.
+            """
+            count_dir_entries = 0
+            direct_entries = dir_path.glob("*")
+            for item_path in direct_entries:
+                count_dir_entries += 1
+                if not item_path.is_dir() and not item_path.is_symlink():
+                    unexpected_dirs.append(controller)
+                    break
+            else:
+                if count_dir_entries == 0:
+                    hier.add_controller_list(Hierarchy.EMPTY_CONTROLLERS, controller)
+                    continue
+
+            if unexpected_dirs:
+                hier.add_controller_list(Hierarchy.UNEXPECTED_CONTROLLERS, controller)
+            hier.add_verify_list(controller)
+
+    return 0
+
+
+# User Hierarchy
+def verify_users(hier):
+    """ 'hier' is UserHierarchy class object."""
+    if not Path(hier.path).is_dir():
+        print(
+            f"The setting for USERS in the config file is {hier.path}, but that is"
+            f" not a directory"
+        )
+        return 1
+
+    user_dirs = Path(hier.path).glob("*")
+
+    for user_path in user_dirs:
+        if user_path.is_dir():
+            hier.add_user_dir(user_path)
+        else:
+            hier.add_unexpected_objects(user_path.name)
+
+    return 0
+
+
+def check_and_dump(f, hier, ihier):
+
+    cnt = 0
+    first_check = hier.check_controller()
+    second_check = ihier.check_controller()
+    if first_check or second_check:
+        hier.header(f, "start")
+        if first_check:
+            cnt = hier.dump(f)
+        if second_check:
+            cnt += ihier.dump(f)
+        hier.header(f, "end")
+    return cnt
+
+
+def check_directory_exists(pbdir_path, pbdirname):
+    """ checks the existence of directories """
+    pbdir = Path(pbdir_path)
+    try:
+        pbdir_p = pbdir.resolve(strict=True)
+    except FileNotFoundError:
+        print(f"{_NAME_}: Bad {pbdirname}={pbdir}", file=sys.stderr)
+        return 1
+
+    if not pbdir_p.is_dir():
+        print(f"{_NAME_}: Bad {pbdirname}={pbdir}", file=sys.stderr)
+        return 1
+
+    return 0
```

Comment:
This function should return proper boolean values (`True`/`False`), and it should be named appropriately for the values that it returns.  (E.g., `directory_exists()` returns `True` when the indicated directory exists.)

Reply:
Using exceptions in this function would be good; however, I'm not sure they would work well in many of the other cases.

From what I've seen, this code has lots of functions which return a yes/no answer as to whether a particular directory entry should be reported. In those cases, I suspect that a "yes" answer is not an "exceptional condition", at least in the context of those functions.

Reply:
Yes, there are distinct classes of two-value utility functions:

1. Is it _this_ or _that_, which is a boolean, and best named as you've suggested elsewhere as `isitthat`;
2. Operations or checks where one answer is expected/normal and the other is unexpected/abnormal, and these are best handled with no return value and a clear exception raised on the abnormal case.

Reply:
done